zcash_test_vectors/unified_full_viewing_keys.py

```diff
@@ -3,45 +3,54 @@
 
 from random import Random
 
-from cryptography.hazmat.primitives.asymmetric import ec
-from cryptography.hazmat.primitives.serialization import PublicFormat, Encoding
-
 from .output import render_args, render_tv, Some
 from .rand import Rand, randbytes
 from .orchard import key_components as orchard_key_components
 from .sapling import zip32 as sapling_zip32
+from .transparent import bip_0032
+from .hd_common import ZCASH_MAIN_COINTYPE, hardened
 from .unified_encoding import encode_unified, decode_unified
 from .unified_encoding import P2PKH_ITEM, SAPLING_ITEM, ORCHARD_ITEM
 
+
 def main():
     args = render_args()
 
     rng = Random(0xabad533d)
     rand = Rand(randbytes(rng))
-    seed = rand.b(32)
+    seed = bytes(range(32))
+
+    t_root_key = bip_0032.ExtendedSecretKey.master(seed)
+    t_purpose_key = t_root_key.child(hardened(44))
+    t_coin_key = t_purpose_key.child(hardened(ZCASH_MAIN_COINTYPE))
+
+    s_root_key = sapling_zip32.ExtendedSpendingKey.master(seed)
+    s_purpose_key = s_root_key.child(hardened(32))
+    s_coin_key = s_purpose_key.child(hardened(ZCASH_MAIN_COINTYPE))
+
+    o_root_key = orchard_key_components.ExtendedSpendingKey.master(seed)
+    o_purpose_key = o_root_key.child(hardened(32))
+    o_coin_key = o_purpose_key.child(hardened(ZCASH_MAIN_COINTYPE))
 
     test_vectors = []
-    for i in range(0, 10):
+    for account in range(0, 20):
         has_t_key = rand.bool()
         if has_t_key:
-            c = rand.b(32)
-            privkey = ec.derive_private_key(int.from_bytes(rand.b(32), 'little'), ec.SECP256K1())
-            pubkey = privkey.public_key()
-            pubkey_bytes = pubkey.public_bytes(Encoding.X962, PublicFormat.CompressedPoint)
-            assert len(pubkey_bytes) == 33
-            assert pubkey_bytes[0] in (0x02, 0x03)
-            t_key_bytes = c + pubkey_bytes
+            rand.b(20) # discard, to match UA generation
+
+            # <https://zips.z.cash/zip-0316#encoding-of-unified-full-incoming-viewing-keys>
+            # "However, the [Transparent P2PKH] FVK uses the key at the Account level, i.e.
+            # at path m/44'/coin_type'/account', while the IVK uses the external (non-change)
+            # child key at the Change level, i.e. at path m/44'/coin_type'/account'/0."
+            t_account_key = t_coin_key.child(hardened(account))
+            t_key_bytes = bytes(t_account_key.public_key())
```

Comment:
Blocking: I believe that this is missing the chain code.

Reply:
Edit: okay, this is not the public key, it's the extended public key, which has the correct byte representation.